### PR TITLE
feat: Code locations for metrics

### PR DESCRIPTION
**Pull request by jan-auer**
DDM wants to show code locations with metrics. Locations are semi-static information: they change infrequently, so they don't need to be reported with every data point.

Sentry expects locations to be reported at least once per day. With backdating of metrics, the timestamp used to report the location is the metric bucket's timestamp rounded down to the start of the day (UTC timezone).

The metrics aggregator keeps a cache of previously reported locations. When a location is seen for the first time on a day, it is added to a list of pending locations. On the next flush cycle, all pending locations are sent to Sentry in the same envelope as the metric buckets.

See: https://github.com/getsentry/relay/pull/2751
Epic: https://github.com/getsentry/sentry/issues/60260

### CODE REVIEW

**Comment by antonpirker** [comment]
The mock imports should be like this to work everywhere:
```
try:
    from unittest import mock  # python 3.3 and above
except ImportError:
    import mock  # python < 3.3
```

File: sentry_sdk/client.py [hunk 1]
```diff
@@ -236,11 +236,13 @@ def _capture_envelope(envelope):
...
             self.metrics_aggregator = None  # type: Optional[MetricsAggregator]
-            if self.options.get("_experiments", {}).get("enable_metrics"):
+            experiments = self.options.get("_experiments", {})
+            if experiments.get("enable_metrics"):
                 from sentry_sdk.metrics import MetricsAggregator
 
                 self.metrics_aggregator = MetricsAggregator(
-                    capture_func=_capture_envelope
+                    capture_func=_capture_envelope,
+                    enable_code_locations=bool(experiments.get("code_locations")),
```
Comment (antonpirker) [inline sentry_sdk/client.py:245]: should be `metric_code_locations`, no?

File: tests/test_metrics.py [hunk 2]
```diff
@@ -40,7 +43,7 @@ def test_incr(sentry_init, capture_envelopes):
     sentry_init(
         release="fun-release",
         environment="not-fun-env",
-        _experiments={"enable_metrics": True},
+        _experiments={"enable_metrics": True, "code_locations": True},
```
Comment (antonpirker) [inline tests/test_metrics.py:46]: Should be `metric_code_locations`.

File: tests/test_metrics.py [hunk 3]
```diff
@@ -66,12 +69,29 @@ def test_incr(sentry_init, capture_envelopes):
...
+        },
+    }
+
 
 def test_timing(sentry_init, capture_envelopes):
     sentry_init(
         release="fun-release@1.0.0",
         environment="not-fun-env",
-        _experiments={"enable_metrics": True},
+        _experiments={"enable_metrics": True, "code_locations": True},
```
Comment (antonpirker) [inline tests/test_metrics.py:94]: Should be `metric_code_locations`.

File: tests/test_metrics.py [hunk 4]
```diff
@@ -97,12 +117,29 @@ def test_timing(sentry_init, capture_envelopes):
...
+        },
+    }
+
 
 def test_timing_decorator(sentry_init, capture_envelopes):
     sentry_init(
         release="fun-release@1.0.0",
         environment="not-fun-env",
-        _experiments={"enable_metrics": True},
+        _experiments={"enable_metrics": True, "code_locations": True},
```
Comment (antonpirker) [inline tests/test_metrics.py:142]: Should be `metric_code_locations`.

File: tests/test_metrics.py [hunk 5]
```diff
@@ -147,12 +184,39 @@ def amazing_nano():
...
+        },
+    }
+
 
 def test_timing_basic(sentry_init, capture_envelopes):
     sentry_init(
         release="fun-release@1.0.0",
         environment="not-fun-env",
-        _experiments={"enable_metrics": True},
+        _experiments={"enable_metrics": True, "code_locations": True},
```
Comment (antonpirker) [inline tests/test_metrics.py:219]: Should be `metric_code_locations`.

File: tests/test_metrics.py [hunk 6]
```diff
@@ -180,12 +245,29 @@ def test_timing_basic(sentry_init, capture_envelopes):
...
+        },
+    }
+
 
 def test_distribution(sentry_init, capture_envelopes):
     sentry_init(
         release="fun-release@1.0.0",
         environment="not-fun-env",
-        _experiments={"enable_metrics": True},
+        _experiments={"enable_metrics": True, "code_locations": True},
```
Comment (antonpirker) [inline tests/test_metrics.py:270]: Should be `metric_code_locations`.

File: tests/test_metrics.py [hunk 7]
```diff
@@ -213,12 +295,29 @@ def test_distribution(sentry_init, capture_envelopes):
...
+        },
+    }
+
 
 def test_set(sentry_init, capture_envelopes):
     sentry_init(
         release="fun-release@1.0.0",
         environment="not-fun-env",
-        _experiments={"enable_metrics": True},
+        _experiments={"enable_metrics": True, "code_locations": True},
```
Comment (antonpirker) [inline tests/test_metrics.py:320]: Should be `metric_code_locations`.

**Review by antonpirker — COMMENTED** [review]
Looks good. I few minor changes I mentioned.

File: sentry_sdk/metrics.py [hunk 8]
```diff
@@ -536,12 +610,13 @@ def incr(
     unit="none",  # type: MeasurementUnit
     tags=None,  # type: Optional[MetricTags]
     timestamp=None,  # type: Optional[Union[float, datetime]]
+    stacklevel=1,  # type: int
```
Comment (Dav1dde) [inline sentry_sdk/metrics.py:613]: Stacklevel default should be 0, then add +1>

**Review by Dav1dde — APPROVED** [review]

**Review by antonpirker — APPROVED** [review]
Looks good! 🚢

**Pull request merged**